### PR TITLE
PP-9017 Publish incoming events to SNS

### PR DESCRIPTION
**Pull request by mrlumbu**
Adds a new action for ledger on recieving an event.
If SNS is enabled, Ledger publishes the event to AWS SNS,
using the provide topic.

If SNS is not enabled then publish should not be tried
and everything should behave as before.

I have introduced an enum to represent available SNS topics
to rest of app - feels nicer and safer than any of the
alternatives.

The publisher itself is very simple - don't think anything
controversial.

At the moment if any exception in SNS publishing will
be swallowed and only logged. This is to make totally sure
this does not break environments where SNS is not currently
provisioned.

Set up emitter and hook in to event processing life
cycle

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,5 @@ target
 bin
 
 .DS_Store
+
+.java-version
```
Comment (alexbishop1) [inline .gitignore:34]: I’m inclined to just check this file in.

File: build-local.sh [hunk 2]
```diff
@@ -4,5 +4,5 @@ set -e
 
 cd "$(dirname "$0")"
 
-mvn -DskipITs clean verify
+mvn -DskipTests clean verify
```
Comment (alexbishop1) [inline build-local.sh:7]: Are you sure?

File: pom.xml [hunk 3]
```diff
@@ -142,6 +142,11 @@
             <artifactId>aws-java-sdk-sqs</artifactId>
             <version>1.12.129</version>
         </dependency>
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>sns</artifactId>
+            <version>2.17.46</version>
```
Comment (alexbishop1) [inline pom.xml:148]: Latest is version 2.17.99.

File: src/main/java/uk/gov/pay/ledger/eventpublisher/EventPublisher.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    public void publishMessageToTopic(String message, TopicName topicName) throws EventPublisherException{
+        var topicArn = topicNameArnMapper.getArnForTopicName(topicName);
+        try {
+            PublishRequest request = PublishRequest.builder()
+                    .message(message)
+                    .topicArn(topicArn)
+                    .build();
+            snsClient.publish(request);
+        } catch (Exception e) {
+            LOGGER.warn("Failed to publish message to SNS");
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/ledger/eventpublisher/EventPublisher.java:31]: We should probably have some indication of what the message is.

File: src/test/java/uk/gov/pay/ledger/queue/EventMessageHandlerTest.java [hunk 5]
```diff
@@ -119,4 +139,28 @@ void shouldScheduleMessageForRetry_WhenEventIsNotProcessedSuccessfully() throws
...
+
+    @Test
+    void shouldPublishToMessageWhenSnsEnabled() throws QueueException {
+        Event event = aQueuePaymentEventFixture().toEntity();
+        when(eventMessage.getEvent()).thenReturn(event);
+
+        eventMessageHandler.handle();
+
+        verifyNoInteractions(eventPublisher);
+    }
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/ledger/queue/EventMessageHandlerTest.java:165]: The names of these tests seem to be the wrong way around!

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
Basically fine.

File: src/main/java/uk/gov/pay/ledger/eventpublisher/EventPublisher.java (hunk 4 above)
Comment (mrlumbu) [reply src/main/java/uk/gov/pay/ledger/eventpublisher/EventPublisher.java:31]: yes, but at this point the message is just a string, which we obvs don't want to log. Could explicitly pass in message id in to method for logging purposes?

Comment (mrlumbu) [reply src/main/java/uk/gov/pay/ledger/eventpublisher/EventPublisher.java]: having discussed, agreed that responsibility for logging should be in caller, EventPublisher just needs to throw sufficiently informative exception

**Review by alexbishop1 — APPROVED** [review]

**Pull request merged**